### PR TITLE
Set upper Numpy Version

### PR DESCRIPTION
**Pull request by stbaione**
# Description

Found a bug when walking through the shortfin llm docs using latest `nightly` sharktank. gguf is currently incompatible with numpy > 2. This breaks `sharktank.examples.export_paged_llm_v1` on linux.

The gguf issue is filed [here](https://github.com/ggerganov/llama.cpp/issues/9021). It was closed from inactivity, but isn't actually solved and has a PR open for the fix.

## Repro Steps
On linux:

### Before re-pinning
Create a virtual environment:
```bash
python -m venv --prompt sharktank .venv
souce .venv/bin/activate
```

Install depencies and sharktank:
```bash
pip install -r pytorch-cpu-requirements.txt
pip install -r requirements.txt -e sharktank/
```

Show numpy version (before re-pinning):
```bash
pip show numpy | grep Version
Version: 2.1.3
```

Try running `export_paged_llm_v1`:
```bash
python -m sharktank.examples.export_paged_llm_v1 --gguf-file=$PATH_TO_GGUF --output-mlir=./temp/model.mlir --output-config=./temp/config.json --bs=1,4
```

You'll see this error:
```text
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/stbaione/repos/SHARK-Platform/sharktank/sharktank/examples/export_paged_llm_v1.py", line 336, in <module>
    main()
  File "/home/stbaione/repos/SHARK-Platform/sharktank/sharktank/examples/export_paged_llm_v1.py", line 67, in main
    dataset = cli.get_input_dataset(args)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stbaione/repos/SHARK-Platform/sharktank/sharktank/utils/cli.py", line 104, in get_input_dataset
    return Dataset.load(data_files["gguf"], file_type="gguf")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stbaione/repos/SHARK-Platform/sharktank/sharktank/types/theta.py", line 347, in load
    ds = _dataset_load_helper(path, file_type=file_type, mmap=mmap)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stbaione/repos/SHARK-Platform/sharktank/sharktank/types/theta.py", line 536, in _dataset_load_helper
    return gguf_interop.load_file(path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stbaione/repos/SHARK-Platform/sharktank/sharktank/types/gguf_interop/base.py", line 117, in load_file
    reader = GGUFReader(gguf_path)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/stbaione/repos/SHARK-Platform/.venv_2/lib/python3.12/site-packages/gguf/gguf_reader.py", line 87, in __init__
    if self._get(offs, np.uint32, override_order = '<')[0] != GGUF_MAGIC:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stbaione/repos/SHARK-Platform/.venv_2/lib/python3.12/site-packages/gguf/gguf_reader.py", line 137, in _get
    .newbyteorder(override_order or self.byte_order)
     ^^^^^^^^^^^^
AttributeError: `newbyteorder` was removed from the ndarray class in NumPy 2.0. Use `arr.view(arr.dtype.newbyteorder(order))` instead.
```

## After re-pinning
Create a virtual environment:
```bash
python -m venv --prompt sharktank .venv
souce .venv/bin/activate
```

Install depencies and sharktank:
```bash
pip install -r pytorch-cpu-requirements.txt
pip install -r requirements.txt -e sharktank/
```

Show numpy version:
```bash
pip show numpy | grep Version
Version: 1.26.3
```

Run `export_paged_llm_v1`:
```bash
python -m sharktank.examples.export_paged_llm_v1 --gguf-file=$PATH_TO_GGUF --output-mlir=./temp/model.mlir --output-config=./temp/config.json --bs=1,4
```

With re-pinning we get desired output:
```text
Exporting decode_bs1
Exporting prefill_bs4
Exporting decode_bs4
GENERATED!
Exporting
Saving to './temp/model.mlir'
```

### CODE REVIEW

File: sharktank/requirements.txt [hunk 1]
```diff
@@ -2,8 +2,7 @@ iree-turbine
 
 # Runtime deps.
 gguf==0.6.0
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/requirements.txt:4]: It doesn't appear so:

- They still have the call to `newbyteorder` in gguf_read.py: https://github.com/ggerganov/llama.cpp/blob/master/gguf-py/gguf/gguf_reader.py#L151

- Their numpy dependency specified >= 1.17, but doesn't have an upper bound: https://github.com/ggerganov/llama.cpp/blob/master/gguf-py/pyproject.toml#L23

[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/requirements.txt:4]: Also appears to still have an active issue in `cumfy-ui`: https://github.com/city96/ComfyUI-GGUF/issues/7 (reopened last week)

File: sharktank/requirements.txt [hunk 2]
```diff
@@ -2,8 +2,7 @@ iree-turbine
...
-numpy==1.26.3; sys_platform == 'win32'
-numpy; sys_platform != 'win32'
+numpy==1.26.3
```
Comment (marbre) [inline sharktank/requirements.txt:5]: Must it be specifically `1.26.3`or does `<2` work as well? Or is `1.26.4` already incompatible?

Comment (stbaione) [reply sharktank/requirements.txt:5]: <2 seems to work as well:

In fresh venv:
```bash
pip install -r pytorch-cpu-requirements.txt
pip install -r requirements.txt -e sharktank
pip uninstall numpy
pip install "numpy<2"
```

```bash
python -m sharktank.examples.export_paged_llm_v1 --gguf-file=./sglang_integration_tests/llama3.1-8b/llama8b_f16.gguf --output-mlir=./temp/model.mlir --output-config=./temp/config.json --bs=1,4
```

Output:
```
Exporting decode_bs1
Exporting prefill_bs4
Exporting decode_bs4
GENERATED!
Exporting
Saving to './temp/model.mlir'
```

Comment (stbaione) [reply sharktank/requirements.txt:5]: Do we wanna leave the version pinned for windows, or set both to `<2`?

Comment (marbre) [reply sharktank/requirements.txt:5]: In that case, let's please not over-strictly pin it to only one specific version and go with `<2.0` or `<2` instead if that#s really needed. However, I assume it's okay to not have it strictly pinned in the CI as well (changed as part of commit 4fcf77b)

Comment (marbre) [reply sharktank/requirements.txt:5]: > Do we wanna leave the version pinned for windows, or set both to `<2`?

Well, I would say let's specify the version range for both. We only need to pin it for `sharktank` not for any use in `shortfin` corret?

Comment (stbaione) [reply sharktank/requirements.txt]: Set `numpy<2.0` for windows and non-windows

File: sharktank/requirements.txt [hunk 3]
```diff
@@ -2,7 +2,8 @@ iree-turbine
 
 # Runtime deps.
 gguf==0.6.0
-numpy==1.26.3
+numpy<2.0; sys_platform == 'win32'
+numpy<2.0; sys_platform != 'win32'
```
Comment (marbre) [inline sharktank/requirements.txt:6]:
```suggestion
numpy<2.0
```

Comment (stbaione) [reply sharktank/requirements.txt]: Updated

**Comment by marbre** [comment]
Maybe rename to `Set upper Numpy version` as we don't strictly pin?

**Comment by stbaione** [comment]
> Maybe rename to `Set upper Numpy version` as we don't strictly pin?

@marbre Updated!

**Review by marbre — APPROVED** [review]

**Pull request merged**